src/lib.rs

```diff
@@ -109,51 +111,58 @@ macro_rules! dataset {
                 translation: Vec<i64>,
                 dimensions: Vec<i64>,
             ) -> PyResult<Py<PyArrayDyn<$d_type>>> {
-                let bounding_box = BoundingBox::new(translation.into(), dimensions.into());
+                let arr = py.allow_threads(move || {
+                    let bounding_box = BoundingBox::new(translation.into(), dimensions.into());
 
-                let arr = self
-                    .n5
-                    .read_ndarray::<$d_type>(&self.path, &self.attr, &bounding_box)?;
+                    self.n5
+                        .read_ndarray::<$d_type>(&self.path, &self.attr, &bounding_box)
+                })?;
                 Ok(arr.into_pyarray(py).to_owned())
             }
 
             fn write_ndarray(
                 &self,
+                py: Python,
                 translation: Vec<i64>,
                 arr: &PyArrayDyn<$d_type>,
                 fill_val: $d_type,
             ) -> PyResult<()> {
-                self.n5.write_ndarray::<$d_type>(
-                    &self.path, &self.attr, translation.into(),
-                    // TODO: because of n5's `write_ndarray` signature, must
-                    // pass a reference to an owned ndarray here. n5 could
-                    // instead take an array view, which may solve this.
-                    &arr.as_array().to_owned(), fill_val)?;
+                py.allow_threads(move || {
```

Comment:
I'm a little unclear on how this works. How do you chose the number of threads to use? And would this make the default behavior of reads and writes to be parallelized?
I'm all for it if it speeds up reads and writes, but I don't know if there are situations where this might not be preferred.

Reply:
This doesn't actually do any threading inside rust (or indeed python). All it does is release the GIL: that is, tell *python* "if you wanted to do some threading, you're fine to do other work while this process is going on", where usually it would have to freeze this process . It means that if you used python's `threading`, `ThreadPoolExecutor`, or `asyncio`, rust doing IO won't block python doing python-y things in a different python thread (which is not the same as a python process as supported by `multiprocessing`, `ProcessPoolExecutor` etc.).

Rust-N5 already has advisory locks for blocks it's working on, so there shouldn't be any issues with trying to read or write over partially-written blocks, so long as the reads and writes are coming from the same machine.